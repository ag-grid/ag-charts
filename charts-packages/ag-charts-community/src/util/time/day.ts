import { CountableTimeInterval } from "./interval";
import { durationDay, durationMinute } from "./duration";

function floor(date: Date) {
    date.setHours(0, 0, 0, 0);
}
function offset(date: Date, days: number) {
    date.setDate(date.getDate() + days);
}
function count(start: Date, end: Date): number {
    const tzMinuteDelta = end.getTimezoneOffset() - start.getTimezoneOffset();
    return (end.getTime() - start.getTime() - tzMinuteDelta * durationMinute) / durationDay;
}
function field(date: Date): number {
    return date.getDate() - 1;
}

export const day = new CountableTimeInterval(floor, offset, count, field);
export default day;
