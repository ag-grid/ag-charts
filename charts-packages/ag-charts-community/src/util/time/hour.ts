import { CountableTimeInterval } from "./interval";
import { durationHour, durationMinute, durationSecond } from "./duration";

function floor(date: Date) {
    date.setTime(date.getTime() - date.getMilliseconds() - date.getSeconds() * durationSecond - date.getMinutes() * durationMinute);
}
function offset(date: Date, hours: number) {
    date.setTime(date.getTime() + hours * durationHour);
}
function count(start: Date, end: Date): number {
    return (end.getTime() - start.getTime()) / durationHour;
}
function field(date: Date): number {
    return date.getHours();
}

export const hour = new CountableTimeInterval(floor, offset, count, field);
export default hour;
