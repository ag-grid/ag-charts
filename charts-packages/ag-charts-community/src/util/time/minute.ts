import { CountableTimeInterval } from "./interval";
import { durationMinute, durationSecond } from "./duration";

function floor(date: Date) {
    date.setTime(date.getTime() - date.getMilliseconds() - date.getSeconds() * durationSecond);
}
function offset(date: Date, minutes: number) {
    date.setTime(date.getTime() + minutes * durationMinute);
}
function count(start: Date, end: Date): number {
    return (end.getTime() - start.getTime()) / durationMinute;
}
function field(date: Date): number {
    return date.getMinutes();
}

export const minute = new CountableTimeInterval(floor, offset, count, field);
export default minute;
