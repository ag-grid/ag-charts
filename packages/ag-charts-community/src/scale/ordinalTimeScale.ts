import type { TimeInterval } from '../util/time/interval';
import { buildFormatter } from '../util/timeFormat';
import { dateToNumber, defaultTimeTickFormat } from '../util/timeFormatDefaults';
import { BandScale } from './bandScale';
import { ContinuousScale } from './continuousScale';
import { Invalidating } from './invalidating';
import { TimeScale } from './timeScale';

export class OrdinalTimeScale extends BandScale<Date, TimeInterval | number> {
    override readonly type = 'ordinal-time';

    static override is(value: any): value is OrdinalTimeScale {
        return value instanceof OrdinalTimeScale;
    }

    @Invalidating
    tickCount = ContinuousScale.defaultTickCount;
    @Invalidating
    minTickCount = 0;
    @Invalidating
    maxTickCount = Infinity;

    @Invalidating
    override interval?: TimeInterval | number = undefined;

    toDomain(d: number): Date {
        return new Date(d);
    }

    protected override index: Map<Date[], number> = new Map<Date[], number>();

    /**
     * Contains unique datums only. Since `{}` is used in place of `Map`
     * for IE11 compatibility, the datums are converted `toString` before
     * the uniqueness check.
     */
    protected override _domain: Date[] = [];
    override set domain(values: Date[]) {
        this.invalid = true;

        const domain: Date[] = [];

        const n = values.length;
        if (n === 0) {
            this._domain = domain;
            return;
        }

        this.index = new Map<Date[], number>();
        const { index } = this;

        const isReversed = values[0] > values[n - 1];

        values.forEach((value, i) => {
            const nextValue = this.toDomain(
                dateToNumber(values[i + 1]) + (isReversed ? 1 : -1) || dateToNumber(value) + (isReversed ? -1 : 1)
            );

            const dateRange = isReversed ? [nextValue, value] : [value, nextValue];

            if (index.get(dateRange) === undefined) {
                index.set(dateRange, domain.push(value) - 1);
            }
        });

        this._domain = domain;
    }
    override get domain(): Date[] {
        return this._domain;
    }

    override ticks(): Date[] {
        this.refresh();

        const [t0, t1] = [dateToNumber(this.domain[0]), dateToNumber(this.domain.at(-1))];

        const start = Math.min(t0, t1);
        const stop = Math.max(t0, t1);

        const { interval, tickCount, minTickCount, maxTickCount } = this;

        let ticks;
        if (interval !== undefined) {
            const [r0, r1] = this.range;
            const availableRange = Math.abs(r1 - r0);
            ticks = TimeScale.getTicksForInterval({ start, stop, interval, availableRange });
        }

        ticks ??= TimeScale.getDefaultTicks({ start, stop, tickCount, minTickCount, maxTickCount });

        // max one tick per band
        const tickPositions = new Set<number>();
        ticks = ticks.filter((tick) => {
            const position = this.convert(tick);
            if (tickPositions.has(position)) {
                return false;
            }
            tickPositions.add(position);
            return true;
        });

        return ticks;
    }

    override convert(d: Date): number {
        if (typeof d === 'number') {
            d = new Date(d);
        }

        if (!(d instanceof Date)) {
            return NaN;
        }

        this.refresh();

        let i;

        for (const [dateRange, index] of this.index.entries()) {
            if (d >= dateRange[0] && d <= dateRange[1]) {
                i = index;
                break;
            }
        }

        if (i === undefined) {
            return NaN;
        }

        const r = this.ordinalRange[i];
        if (r === undefined) {
            return NaN;
        }

        return r;
    }

    /**
     * Returns a time format function suitable for displaying tick values.
     * @param specifier If the specifier string is provided, this method is equivalent to
     * the {@link TimeLocaleObject.format} method.
     * If no specifier is provided, this method returns the default time format function.
     */
    tickFormat({ ticks, specifier }: { ticks?: any[]; specifier?: string }): (date: Date) => string {
        return specifier == null ? defaultTimeTickFormat(ticks) : buildFormatter(specifier);
    }

    override invert(y: number): Date {
        return new Date(super.invert(y));
    }

    override invertNearest(y: number): Date {
        return new Date(super.invertNearest(y));
    }
}
