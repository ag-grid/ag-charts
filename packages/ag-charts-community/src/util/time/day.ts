import { CountableTimeInterval } from './interval';
import { durationDay } from './duration';

function encode(date: Date) {
    const tzOffsetMs = date.getTimezoneOffset() * 60_000;

    return Math.floor((date.getTime() - tzOffsetMs) / durationDay);
}

function decode(encoded: number) {
    const d = new Date(0);
    const startTzOffsetMs = d.getTimezoneOffset() * 60_000;
    d.setTime(d.getTime() + startTzOffsetMs);
    d.setDate(d.getDate() + encoded);

    return d;
}

export const day = new CountableTimeInterval(encode, decode);
export default day;
