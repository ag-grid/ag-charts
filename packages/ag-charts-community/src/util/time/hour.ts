import { durationHour, durationMinute } from './duration';
import { CountableTimeInterval } from './interval';

const offset = new Date().getTimezoneOffset() * durationMinute;

function encode(date: Date) {
    return Math.floor((date.getTime() - offset) / durationHour);
}

function decode(encoded: number) {
    return new Date(offset + encoded * durationHour);
}

export const hour = new CountableTimeInterval(encode, decode);
export default hour;
