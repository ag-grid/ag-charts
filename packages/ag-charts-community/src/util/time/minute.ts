import { CountableTimeInterval } from './interval';
import { durationMinute } from './duration';

function floor(date: Date) {
    date.setSeconds(0, 0);
}
function offset(date: Date, minutes: number) {
    date.setMinutes(date.getMinutes() + minutes);
}
function stepTest(date: Date, minutes: number) {
    return Math.floor(date.getTime() / durationMinute) % minutes === 0;
}

export const minute = new CountableTimeInterval(floor, offset, stepTest);
export default minute;
