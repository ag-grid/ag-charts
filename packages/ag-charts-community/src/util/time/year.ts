import { CountableTimeInterval } from './interval';

function encode(date: Date) {
    return date.getFullYear();
}

function decode(encoded: number) {
    // Note: assigning years through the constructor
    // will break for years 0 - 99 AD (will turn 1900's).
    const d = new Date();
    d.setFullYear(encoded);
    d.setMonth(0, 1);
    d.setHours(0, 0, 0, 0);
    return d;
}

export const year = new CountableTimeInterval(encode, decode);
export default year;
