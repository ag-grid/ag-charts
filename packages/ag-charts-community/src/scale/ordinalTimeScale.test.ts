import { describe, expect, it } from '@jest/globals';

import day from '../util/time/day';
import {
    durationDay,
    durationHour,
    durationMinute,
    durationMonth,
    durationWeek,
    durationYear,
} from '../util/time/duration';
import hour from '../util/time/hour';
import minute from '../util/time/minute';
import month from '../util/time/month';
import year from '../util/time/year';
import { OrdinalTimeScale } from './ordinalTimeScale';

describe('OrdinalTimeScale', () => {
    it('should create nice ticks', () => {
        const scale = new OrdinalTimeScale();
        scale.domain = [
            new Date(2024, 1, 26),
            new Date(2024, 1, 27),
            new Date(2024, 1, 28),
            new Date(2024, 1, 29),
            new Date(2024, 2, 1),
            new Date(2024, 2, 4),
            new Date(2024, 2, 5),
            new Date(2024, 2, 6),
        ];
        scale.tickCount = 5;
        expect(scale.ticks()).toMatchSnapshot();
    });

    it('should create ticks matching the data domain if the domain length is smaller than maxTickCount', () => {
        const scale = new OrdinalTimeScale();
        scale.domain = [
            new Date(2024, 1, 26),
            new Date(2024, 1, 27),
            new Date(2024, 1, 28),
            new Date(2024, 1, 29),
            new Date(2024, 2, 1),
            new Date(2024, 2, 4),
            new Date(2024, 2, 5),
            new Date(2024, 2, 6),
        ];
        scale.maxTickCount = 8;
        expect(scale.ticks()).toMatchSnapshot();
    });

    it('should extend the domain to create nice ticks matching the data domain', () => {
        const scale = new OrdinalTimeScale();
        scale.domain = [
            // yearly data but not on the first day of the year (January 1st)
            new Date(2023, 3, 5),
            new Date(2024, 3, 5),
            new Date(2025, 3, 5),
        ];
        expect(scale.ticks()).toMatchSnapshot();
    });

    describe('should create ticks with configured', () => {
        describe(`milliseconds interval`, () => {
            const MILLISECONDS_INTERVALS = [
                {
                    name: 'every minute',
                    interval: durationMinute * 5, // tick every 5 mins
                    domain: [
                        // Datum every minute
                        new Date(2024, 0, 1, 12, 0),
                        new Date(2024, 0, 1, 12, 1),
                        new Date(2024, 0, 1, 12, 2),
                        new Date(2024, 0, 1, 12, 3),
                        new Date(2024, 0, 1, 12, 4),
                        new Date(2024, 0, 1, 12, 5),
                        new Date(2024, 0, 1, 12, 6),
                        new Date(2024, 0, 1, 12, 7),
                        new Date(2024, 0, 1, 12, 8),
                        new Date(2024, 0, 1, 12, 9),
                        new Date(2024, 0, 1, 12, 10),
                        new Date(2024, 0, 1, 12, 11),
                        new Date(2024, 0, 1, 12, 12),
                        new Date(2024, 0, 1, 12, 13),
                        new Date(2024, 0, 1, 12, 14),
                        new Date(2024, 0, 1, 12, 15),
                    ],
                },
                {
                    name: 'every hour',
                    interval: durationHour,
                    domain: [
                        // datum every 30 mins, tick every hour
                        new Date(2024, 0, 1, 12, 0),
                        new Date(2024, 0, 1, 12, 30),
                        // gap between 12:30PM and 2PM
                        new Date(2024, 0, 1, 14, 0),
                        new Date(2024, 0, 1, 14, 30),
                        new Date(2024, 0, 1, 15, 0),
                        new Date(2024, 0, 1, 15, 30),
                        new Date(2024, 0, 1, 16, 0),
                        new Date(2024, 0, 1, 16, 30),
                        new Date(2024, 0, 1, 17, 0),
                        new Date(2024, 0, 1, 17, 30),
                        new Date(2024, 0, 1, 18, 0),
                        new Date(2024, 0, 1, 18, 30),
                        // next day from 7AM
                        new Date(2024, 0, 2, 7, 0),
                        new Date(2024, 0, 2, 7, 30),
                        new Date(2024, 0, 2, 8, 0),
                        new Date(2024, 0, 2, 8, 30),
                    ],
                },
                {
                    name: 'every half day',
                    interval: durationDay / 2,
                    domain: [
                        // datum every 30 mins spanning 24 hours, tick every half day
                        new Date(2024, 0, 1, 12, 0),
                        new Date(2024, 0, 1, 12, 30),
                        // gap between 12:30PM and 2PM
                        new Date(2024, 0, 1, 14, 0),
                        new Date(2024, 0, 1, 14, 30),
                        new Date(2024, 0, 1, 15, 0),
                        new Date(2024, 0, 1, 15, 30),
                        new Date(2024, 0, 1, 16, 0),
                        new Date(2024, 0, 1, 16, 30),
                        new Date(2024, 0, 1, 17, 0),
                        new Date(2024, 0, 1, 17, 30),
                        new Date(2024, 0, 1, 18, 0),
                        new Date(2024, 0, 1, 18, 30),
                        // next day from 7AM
                        new Date(2024, 0, 2, 7, 0),
                        new Date(2024, 0, 2, 7, 30),
                        new Date(2024, 0, 2, 8, 0),
                        new Date(2024, 0, 2, 8, 30),
                        new Date(2024, 0, 2, 9, 0),
                        new Date(2024, 0, 2, 9, 30),
                        new Date(2024, 0, 2, 10, 0),
                        new Date(2024, 0, 2, 10, 30),
                        new Date(2024, 0, 2, 11, 0),
                        new Date(2024, 0, 2, 11, 30),
                        new Date(2024, 0, 2, 12, 0),
                    ],
                },
                {
                    name: 'every 2 days',
                    interval: durationDay * 2,
                    domain: [
                        new Date(2024, 1, 26), // Mon
                        new Date(2024, 1, 27), // Tue
                        new Date(2024, 1, 28), // Wed
                        new Date(2024, 1, 29), // Thur
                        new Date(2024, 2, 1), // Fri
                        // no weekend data
                        new Date(2024, 2, 4), // Mon
                        new Date(2024, 2, 5), // Tue
                        new Date(2024, 2, 6), // Wed
                    ],
                },
                {
                    name: 'every 2 weeks',
                    interval: durationWeek * 2,
                    domain: [
                        new Date(2024, 1, 5), // week 1
                        new Date(2024, 1, 6),
                        new Date(2024, 1, 7),
                        new Date(2024, 1, 8),
                        new Date(2024, 1, 9),
                        new Date(2024, 1, 12), // week 2
                        new Date(2024, 1, 13),
                        new Date(2024, 1, 14),
                        new Date(2024, 1, 15),
                        new Date(2024, 1, 16),
                        new Date(2024, 1, 19), // week 3
                        new Date(2024, 1, 20),
                        new Date(2024, 1, 21),
                        new Date(2024, 1, 22),
                        new Date(2024, 1, 23),
                        new Date(2024, 1, 26), // week 4
                        new Date(2024, 1, 27),
                        new Date(2024, 1, 28),
                        new Date(2024, 1, 29),
                        new Date(2024, 2, 4),
                    ],
                },
                {
                    name: 'every month',
                    interval: durationMonth,
                    domain: [
                        new Date(2024, 0, 15),
                        new Date(2024, 0, 22),
                        new Date(2024, 0, 29),
                        new Date(2024, 1, 5),
                        new Date(2024, 1, 12),
                        new Date(2024, 1, 19),
                        new Date(2024, 1, 26),
                    ],
                },
                {
                    name: 'every year',
                    interval: durationYear,
                    domain: [
                        new Date(2020, 0, 1),
                        new Date(2020, 3, 1),
                        new Date(2020, 6, 1),
                        new Date(2020, 9, 1),
                        new Date(2020, 12, 1),

                        new Date(2021, 0, 1),
                        new Date(2021, 3, 1),
                        new Date(2021, 6, 1),
                        new Date(2021, 9, 1),
                        new Date(2021, 12, 1),

                        new Date(2022, 0, 1),
                        new Date(2022, 3, 1),
                        new Date(2022, 6, 1),
                        new Date(2022, 9, 1),
                        new Date(2022, 12, 1),

                        new Date(2023, 0, 1),
                        new Date(2023, 3, 1),
                        new Date(2023, 6, 1),
                        new Date(2023, 9, 1),
                        new Date(2023, 12, 1),
                    ],
                },
            ];

            it.each(MILLISECONDS_INTERVALS)(`for $name case`, ({ interval, domain }) => {
                const scale = new OrdinalTimeScale();

                scale.range = [0, 600];
                scale.domain = domain;
                scale.interval = interval;

                expect(scale.ticks()).toMatchSnapshot();
            });
        });

        describe(`time interval`, () => {
            const TIME_INTERVALS = [
                {
                    name: 'every minute',
                    interval: minute,
                    domain: [
                        // Datum every 20 seconds
                        new Date(2024, 0, 1, 17, 53, 0),
                        new Date(2024, 0, 1, 17, 53, 20),
                        new Date(2024, 0, 1, 17, 53, 40),
                        new Date(2024, 0, 1, 17, 54, 0),
                        new Date(2024, 0, 1, 17, 54, 20),
                        new Date(2024, 0, 1, 17, 54, 40),
                        new Date(2024, 0, 1, 17, 55, 0),
                        new Date(2024, 0, 1, 17, 55, 20),
                        new Date(2024, 0, 1, 17, 55, 40),
                        new Date(2024, 0, 1, 17, 56, 0),
                        new Date(2024, 0, 1, 17, 56, 20),
                        new Date(2024, 0, 1, 17, 56, 40),
                        new Date(2024, 0, 1, 17, 57, 0),
                        new Date(2024, 0, 1, 17, 57, 20),
                        new Date(2024, 0, 1, 17, 57, 40),
                        new Date(2024, 0, 1, 17, 58, 0),
                        new Date(2024, 0, 1, 17, 58, 20),
                        new Date(2024, 0, 1, 17, 58, 40),
                        new Date(2024, 0, 1, 17, 59, 0),
                        new Date(2024, 0, 1, 17, 59, 20),
                        new Date(2024, 0, 1, 17, 59, 40),
                    ],
                },
                {
                    name: 'every hour',
                    interval: hour,
                    domain: [
                        // datum every 15 mins, tick every hour
                        new Date(2024, 0, 1, 11, 0),
                        new Date(2024, 0, 1, 11, 15),
                        new Date(2024, 0, 1, 11, 30),
                        new Date(2024, 0, 1, 11, 45),
                        new Date(2024, 0, 1, 12, 0),
                        // gap between 12PM and 4PM
                        new Date(2024, 0, 1, 16, 0),
                        new Date(2024, 0, 1, 16, 15),
                        new Date(2024, 0, 1, 16, 30),
                        new Date(2024, 0, 1, 16, 45),
                        new Date(2024, 0, 1, 17, 0),
                        new Date(2024, 0, 1, 17, 15),
                        new Date(2024, 0, 1, 17, 30),
                        new Date(2024, 0, 1, 17, 45),
                        new Date(2024, 0, 1, 18, 0),
                    ],
                },
                {
                    name: 'every day',
                    interval: day,
                    domain: [
                        // datum every hour spanning 3 days, tick every day
                        new Date(2024, 0, 1, 14, 0),
                        new Date(2024, 0, 1, 15, 0),
                        new Date(2024, 0, 1, 16, 0),
                        new Date(2024, 0, 1, 17, 0),
                        new Date(2024, 0, 1, 18, 0),
                        // next day from 7AM
                        new Date(2024, 0, 2, 7, 0),
                        new Date(2024, 0, 2, 8, 0),
                        new Date(2024, 0, 2, 9, 0),
                        new Date(2024, 0, 2, 10, 0),
                        new Date(2024, 0, 2, 11, 0),
                        new Date(2024, 0, 2, 12, 0),
                        new Date(2024, 0, 2, 13, 0),
                        new Date(2024, 0, 2, 14, 0),
                        new Date(2024, 0, 2, 15, 0),
                        new Date(2024, 0, 2, 16, 0),
                        new Date(2024, 0, 2, 17, 0),
                        new Date(2024, 0, 2, 18, 0),
                        // next day from 7 AM
                        new Date(2024, 0, 3, 7, 0),
                        new Date(2024, 0, 3, 8, 0),
                        new Date(2024, 0, 3, 9, 0),
                        new Date(2024, 0, 3, 10, 0),
                        new Date(2024, 0, 3, 11, 0),
                        new Date(2024, 0, 3, 12, 0),
                        new Date(2024, 0, 3, 13, 0),
                        new Date(2024, 0, 3, 14, 0),
                        new Date(2024, 0, 3, 15, 0),
                        new Date(2024, 0, 3, 16, 0),
                        new Date(2024, 0, 3, 17, 0),
                        new Date(2024, 0, 3, 18, 0),
                    ],
                },
                {
                    name: 'every 3 days',
                    interval: day.every(3),
                    domain: [
                        // datum twice a day
                        new Date(2024, 0, 1, 7, 0),
                        new Date(2024, 0, 1, 18, 0),
                        new Date(2024, 0, 2, 7, 0),
                        new Date(2024, 0, 2, 18, 0),
                        new Date(2024, 0, 3, 7, 0),
                        new Date(2024, 0, 3, 18, 0),
                        new Date(2024, 0, 4, 7, 0),
                        new Date(2024, 0, 4, 18, 0),
                        new Date(2024, 0, 5, 7, 0),
                        new Date(2024, 0, 5, 18, 0),
                        // weekend gap, start again on Monday
                        new Date(2024, 0, 8, 7, 0),
                        new Date(2024, 0, 8, 18, 0),
                        new Date(2024, 0, 9, 7, 0),
                        new Date(2024, 0, 9, 18, 0),
                        new Date(2024, 0, 10, 7, 0),
                        new Date(2024, 0, 10, 18, 0),
                        new Date(2024, 0, 11, 7, 0),
                        new Date(2024, 0, 11, 18, 0),
                        new Date(2024, 0, 12, 7, 0),
                        new Date(2024, 0, 12, 18, 0),
                    ],
                },
                {
                    name: 'every month',
                    interval: month,
                    domain: [
                        new Date(2024, 0, 1),
                        new Date(2024, 0, 8),
                        new Date(2024, 0, 15),
                        new Date(2024, 0, 22),
                        new Date(2024, 0, 29),
                        new Date(2024, 1, 5),
                        new Date(2024, 1, 12),
                        new Date(2024, 1, 19),
                        new Date(2024, 1, 26),
                        new Date(2024, 2, 4),
                    ],
                },
                {
                    name: 'every 2 months',
                    interval: month.every(2),
                    domain: [
                        // datum every 15th of the month spanning across two years
                        new Date(2023, 0, 15),
                        new Date(2023, 1, 15),
                        new Date(2023, 2, 15),
                        new Date(2023, 3, 15),
                        new Date(2023, 4, 15),
                        new Date(2023, 5, 15),
                        new Date(2023, 6, 15),
                        new Date(2023, 7, 15),
                        new Date(2023, 8, 15),
                        new Date(2023, 9, 15),
                        new Date(2023, 10, 15),
                        new Date(2023, 11, 15),
                        new Date(2023, 12, 15),
                        new Date(2024, 0, 15),
                        new Date(2024, 1, 15),
                    ],
                },
                {
                    name: 'every 6 months',
                    interval: month.every(6),
                    domain: [
                        // datum every month spanning across two years
                        new Date(2023, 0, 0),
                        new Date(2023, 1, 0),
                        new Date(2023, 2, 0),
                        new Date(2023, 3, 0),
                        new Date(2023, 4, 0),
                        new Date(2023, 5, 0),
                        new Date(2023, 6, 0),
                        new Date(2023, 7, 0),
                        new Date(2023, 8, 0),
                        new Date(2023, 9, 0),
                        new Date(2023, 10, 0),
                        new Date(2023, 11, 0),
                        new Date(2023, 12, 0),
                        new Date(2024, 0, 0),
                        new Date(2024, 1, 0),
                    ],
                },
                {
                    name: 'every year',
                    interval: year,
                    domain: [
                        // datum every year
                        new Date(2018, 0, 0),
                        new Date(2019, 0, 0),
                        new Date(2020, 0, 0),
                        // no datum for 2021
                        new Date(2022, 0, 0),
                        new Date(2023, 0, 0),
                        new Date(2024, 0, 0),
                    ],
                },
                {
                    name: 'every 3 year',
                    interval: year.every(3),
                    domain: [
                        // datum every 6 months
                        new Date(2018, 0, 0),
                        new Date(2018, 5, 0),
                        new Date(2019, 0, 0),
                        new Date(2019, 5, 0),
                        new Date(2020, 0, 0),
                        new Date(2020, 5, 0),
                        // no datum for 2021
                        new Date(2022, 0, 0),
                        new Date(2022, 5, 0),
                        new Date(2023, 0, 0),
                        new Date(2023, 5, 0),
                        new Date(2024, 0, 0),
                    ],
                },
            ];

            it.each(TIME_INTERVALS)(`for $name case`, ({ interval, domain }) => {
                const scale = new OrdinalTimeScale();

                scale.range = [0, 600];
                scale.domain = domain;
                scale.interval = interval;

                expect(scale.ticks()).toMatchSnapshot();
            });
        });
    });
});
