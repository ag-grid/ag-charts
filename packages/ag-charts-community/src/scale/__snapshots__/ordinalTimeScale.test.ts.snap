// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrdinalTimeScale should create nice ticks 1`] = `
[
  2024-02-27T00:00:00.000Z,
  2024-02-29T00:00:00.000Z,
  2024-03-04T00:00:00.000Z,
  2024-03-06T00:00:00.000Z,
]
`;

exports[`OrdinalTimeScale should create nice ticks when reversed 1`] = `
[
  2024-03-06T00:00:00.000Z,
  2024-03-02T00:00:00.000Z,
  2024-02-29T00:00:00.000Z,
  2024-02-27T00:00:00.000Z,
]
`;

exports[`OrdinalTimeScale should create ticks matching the data domain if the domain length is smaller than maxTickCount 1`] = `
[
  2024-02-26T00:00:00.000Z,
  2024-02-27T00:00:00.000Z,
  2024-02-28T00:00:00.000Z,
  2024-02-29T00:00:00.000Z,
  2024-03-01T00:00:00.000Z,
  2024-03-04T00:00:00.000Z,
  2024-03-05T00:00:00.000Z,
  2024-03-06T00:00:00.000Z,
]
`;

exports[`OrdinalTimeScale should create ticks with configured milliseconds interval for every 2 days case 1`] = `
[
  2024-02-26T00:00:00.000Z,
  2024-02-28T00:00:00.000Z,
  2024-03-01T00:00:00.000Z,
  2024-03-03T00:00:00.000Z,
  2024-03-05T00:00:00.000Z,
]
`;

exports[`OrdinalTimeScale should create ticks with configured milliseconds interval for every 2 weeks case 1`] = `
[
  2024-02-18T00:00:00.000Z,
  2024-03-03T00:00:00.000Z,
]
`;

exports[`OrdinalTimeScale should create ticks with configured milliseconds interval for every half day case 1`] = `
[
  2024-01-01T12:00:00.000Z,
  2024-01-02T00:00:00.000Z,
  2024-01-02T12:00:00.000Z,
]
`;

exports[`OrdinalTimeScale should create ticks with configured milliseconds interval for every hour case 1`] = `
[
  2024-01-01T12:00:00.000Z,
  2024-01-01T13:00:00.000Z,
  2024-01-01T15:00:00.000Z,
  2024-01-01T16:00:00.000Z,
  2024-01-01T17:00:00.000Z,
  2024-01-01T18:00:00.000Z,
  2024-01-01T19:00:00.000Z,
  2024-01-02T08:00:00.000Z,
]
`;

exports[`OrdinalTimeScale should create ticks with configured milliseconds interval for every minute case 1`] = `
[
  2024-01-01T12:00:00.000Z,
  2024-01-01T12:05:00.000Z,
  2024-01-01T12:10:00.000Z,
  2024-01-01T12:15:00.000Z,
]
`;

exports[`OrdinalTimeScale should create ticks with configured milliseconds interval for every month case 1`] = `
[
  2024-02-01T00:00:00.000Z,
]
`;

exports[`OrdinalTimeScale should create ticks with configured milliseconds interval for every year case 1`] = `
[
  2020-01-01T00:00:00.000Z,
  2021-01-01T00:00:00.000Z,
  2022-01-01T00:00:00.000Z,
  2023-01-01T00:00:00.000Z,
  2024-01-01T00:00:00.000Z,
]
`;

exports[`OrdinalTimeScale should create ticks with configured time interval for every 2 months case 1`] = `
[
  2023-03-01T00:00:00.000Z,
  2023-04-30T23:00:00.000Z,
  2023-06-30T23:00:00.000Z,
  2023-08-31T23:00:00.000Z,
  2023-11-01T00:00:00.000Z,
  2024-01-01T00:00:00.000Z,
]
`;

exports[`OrdinalTimeScale should create ticks with configured time interval for every 3 days case 1`] = `
[
  2024-01-04T00:00:00.000Z,
  2024-01-07T00:00:00.000Z,
  2024-01-10T00:00:00.000Z,
]
`;

exports[`OrdinalTimeScale should create ticks with configured time interval for every 3 year case 1`] = `
[
  2020-01-01T00:00:00.000Z,
  2023-01-01T00:00:00.000Z,
]
`;

exports[`OrdinalTimeScale should create ticks with configured time interval for every 6 months case 1`] = `
[
  2023-05-31T23:00:00.000Z,
  2023-12-01T00:00:00.000Z,
]
`;

exports[`OrdinalTimeScale should create ticks with configured time interval for every day case 1`] = `
[
  2024-01-02T00:00:00.000Z,
  2024-01-03T00:00:00.000Z,
]
`;

exports[`OrdinalTimeScale should create ticks with configured time interval for every hour case 1`] = `
[
  2024-01-01T11:00:00.000Z,
  2024-01-01T12:00:00.000Z,
  2024-01-01T13:00:00.000Z,
  2024-01-01T17:00:00.000Z,
  2024-01-01T18:00:00.000Z,
]
`;

exports[`OrdinalTimeScale should create ticks with configured time interval for every minute case 1`] = `
[
  2024-01-01T17:53:00.000Z,
  2024-01-01T17:54:00.000Z,
  2024-01-01T17:55:00.000Z,
  2024-01-01T17:56:00.000Z,
  2024-01-01T17:57:00.000Z,
  2024-01-01T17:58:00.000Z,
  2024-01-01T17:59:00.000Z,
]
`;

exports[`OrdinalTimeScale should create ticks with configured time interval for every month case 1`] = `
[
  2024-01-01T00:00:00.000Z,
  2024-02-01T00:00:00.000Z,
  2024-03-01T00:00:00.000Z,
]
`;

exports[`OrdinalTimeScale should create ticks with configured time interval for every year case 1`] = `
[
  2018-01-01T00:00:00.000Z,
  2019-01-01T00:00:00.000Z,
  2020-01-01T00:00:00.000Z,
  2022-01-01T00:00:00.000Z,
  2023-01-01T00:00:00.000Z,
]
`;

exports[`OrdinalTimeScale should extend the domain to create nice ticks matching the data domain 1`] = `
[
  2023-01-01T00:00:00.000Z,
  2024-01-01T00:00:00.000Z,
  2025-01-01T00:00:00.000Z,
]
`;
